rust/src/operations/transaction/state.rs

```diff
@@ -0,0 +1,335 @@
+use datafusion_common::scalar::ScalarValue;
+use datafusion_common::{Column, DFSchema, Result as DFResult, TableReference};
+use datafusion_expr::{AggregateUDF, Expr, ScalarUDF, TableSource};
+use datafusion_sql::planner::{ContextProvider, SqlToRel};
+use itertools::Either;
+use sqlparser::dialect::GenericDialect;
+use sqlparser::parser::Parser;
+use sqlparser::tokenizer::Tokenizer;
+
+use crate::action::Add;
+use crate::delta_datafusion::to_correct_scalar_value;
+use crate::table_state::DeltaTableState;
+use crate::DeltaResult;
+use crate::{schema, DeltaTableError};
+
+impl DeltaTableState {
+    /// Get the table schema as an [`ArrowSchemaRef`]
+    pub fn arrow_schema(&self) -> DeltaResult<ArrowSchemaRef> {
+        Ok(Arc::new(
+            <ArrowSchema as TryFrom<&schema::Schema>>::try_from(
+                self.schema().ok_or(DeltaTableError::NoMetadata)?,
+            )?,
+        ))
+    }
+
+    /// Iterate over all files in the log matching a predicate
+    pub fn files_matching_predicate(
+        &self,
+        filters: &[Expr],
+    ) -> DeltaResult<impl Iterator<Item = &Add>> {
+        if let Some(Some(predicate)) =
+            (!filters.is_empty()).then_some(conjunction(filters.iter().cloned()))
+        {
+            let pruning_predicate = PruningPredicate::try_new(predicate, self.arrow_schema()?)?;
+            Ok(Either::Left(
+                self.files()
+                    .iter()
+                    .zip(pruning_predicate.prune(self)?.into_iter())
+                    .filter_map(
+                        |(action, keep_file)| {
+                            if keep_file {
+                                Some(action)
+                            } else {
+                                None
+                            }
+                        },
+                    ),
+            ))
+        } else {
+            Ok(Either::Right(self.files().iter()))
+        }
+    }
+
+    /// Parse an expression string into a datafusion [`Expr`]
+    pub fn parse_predicate_expression(&self, expr: impl AsRef<str>) -> DeltaResult<Expr> {
+        let dialect = &GenericDialect {};
+        let mut tokenizer = Tokenizer::new(dialect, expr.as_ref());
+        let tokens = tokenizer
+            .tokenize()
+            .map_err(|err| DeltaTableError::GenericError {
+                source: Box::new(err),
+            })?;
+        let sql = Parser::new(dialect)
+            .with_tokens(tokens)
+            .parse_expr()
+            .map_err(|err| DeltaTableError::GenericError {
+                source: Box::new(err),
+            })?;
+
+        // TODO should we add the table name as qualifier when available?
+        let df_schema = DFSchema::try_from_qualified_schema("", self.arrow_schema()?.as_ref())?;
+        let context_provider = DummyContextProvider::default();
+        let sql_to_rel = SqlToRel::new(&context_provider);
+
+        Ok(sql_to_rel.sql_to_expr(sql, &df_schema, &mut Default::default())?)
+    }
+}
```

Comment:
Parsing predicate expressions is actually kind of broken. Regardless of the schema we pass in, int fields are always recognized as i64. I hope, that the abstractions in the core project will eventually help us out. Until then I plan to either do a follow-up in re-writing these expressions, or, if this is a bug, addressing this upstream in datafusion.